lib/proxy/peer/quicserver.go

```diff
@@ -78,31 +91,357 @@ func (c *QUICServerConfig) checkAndSetDefaults() error {
+			return
+		}
+
+		s.wg.Add(1)
+		go s.handleStream(st, c, log)
+	}
+}
+
+func (s *QUICServer) handleStream(st quic.Stream, c quic.EarlyConnection, log *slog.Logger) {
+	defer s.wg.Done()
+
+	log = log.With("stream_id", st.StreamID())
+	defer log.DebugContext(c.Context(), "Done handling stream.")
+
+	defer st.CancelRead(0)
+	defer st.CancelWrite(0)
+
+	log.DebugContext(c.Context(), "Handling stream.")
+
+	sendErr := func(toSend error) {
+		st.CancelRead(0)
+		defer st.CancelWrite(0)
+		errBuf, err := proto.Marshal(&quicpeeringv1a.DialResponse{
+			Status: status.Convert(trail.ToGRPC(toSend)).Proto(),
+		})
+		if err != nil {
+			return
+		}
+		st.SetWriteDeadline(time.Now().Add(10 * time.Second))
+		if _, err := st.Write(binary.LittleEndian.AppendUint32(nil, uint32(len(errBuf)))); err != nil {
+			return
+		}
+		if _, err := st.Write(errBuf); err != nil {
+			return
+		}
+		if err := st.Close(); err != nil {
+			return
+		}
+	}
+
+	st.SetReadDeadline(time.Now().Add(10 * time.Second))
+	var reqLen uint32
+	if err := binary.Read(st, binary.LittleEndian, &reqLen); err != nil {
+		log.DebugContext(c.Context(), "Failed to read request size.", "error", err)
+		return
+	}
+	if reqLen >= quicMaxMessageSize {
+		log.WarnContext(c.Context(), "Received oversized request.", "request_len", reqLen)
+		return
+	}
+	reqBuf := make([]byte, reqLen)
+	if _, err := io.ReadFull(st, reqBuf); err != nil {
+		log.DebugContext(c.Context(), "Failed to read request.", "error", err)
+		return
+	}
+	st.SetReadDeadline(time.Time{})
+
+	req := new(quicpeeringv1a.DialRequest)
+	if err := proto.Unmarshal(reqBuf, req); err != nil {
+		log.WarnContext(c.Context(), "Failed to unmarshal request.", "error", err)
+		return
+	}
+
+	if req.GetTimestamp() != nil {
+		if time.Since(req.GetTimestamp().AsTime()).Abs() > quicTimestampGraceWindow {
+			log.WarnContext(c.Context(), "0-RTT request has out of sync timestamp, requiring full handshake.",
+				"request_timestamp", req.GetTimestamp().AsTime(),
+			)
+			select {
+			case <-c.HandshakeComplete():
+			case <-c.Context().Done():
+				return
+			}
+		}
+	} else {
+		select {
+		case <-c.HandshakeComplete():
+		case <-c.Context().Done():
+			return
+		default:
+			log.WarnContext(c.Context(), "0-RTT request is missing timestamp, requiring full handshake.")
+			select {
+			case <-c.HandshakeComplete():
+			case <-c.Context().Done():
+				return
+			}
+		}
+	}
+	if req.GetNonce() != 0 {
+		// a replayed request is always wrong even after a full handshake
+		if !s.replayStore.add(req.GetNonce(), time.Now()) {
+			log.ErrorContext(c.Context(), "Request is reusing a nonce, rejecting.", "nonce", req.GetNonce())
+			sendErr(trace.BadParameter("reused or invalid nonce"))
+			return
+		}
+	} else {
+		select {
+		case <-c.HandshakeComplete():
+		case <-c.Context().Done():
+			return
+		default:
+			log.WarnContext(c.Context(), "0-RTT request is missing nonce, requiring full handshake.")
```

Comment:
A client implementation that uses connection pooling doesn't have much need for 0-RTT, so the way that the server is written allows for full handshook(?) connections to deliberately pretend that the replay protection doesn't exist.

Does that make sense? Should we just always require a timestamp and a nonce no matter what, and treat missing values as an error?

Reply:
That makes sense.  If there is a legitimate usecase for not including the nonce, thats fine I think.  Though if it is an intended usecase, I think a warn log might not be warranted, and we probably want the docs on the `nonce` field to explicitly call out that `nonce` is optional and that omission disables 0rtt behavior.

Reply:
I made it so that at least for this "v1alpha" version of the API we're just going to require replay protection no matter what, it makes the server's job much simpler.